Implementation & Unit Tests

Syrus666 opened the pull request:
Uses Fody to weave Classes, Constructors and Methods where .Subscribe() is called on observables.
The returned IDisposable is added to a CompositeDisposable which in turn is disposed when the obj is.
- Sln set up
- Added Obleak.Fody project - The weaver
- Added Obleak.Fody.Core - Attribute and HandleWith extension method for IDisposables
- Unit tests with Class, Method and Constructor level attribute
- Included a static local weaver which can be used for debugging the weaving process
- Set up CI with AppVeyor and auto publishing of NuGet package (currently to MyGet until this gets reviewed)

Commit:
Implementation & Unit Tests
Uses Fody to weave Classes, Constructors and Methods where .Subscribe() is called on observables.
The returned IDisposable is added to a CompositeDisposable which in turn is disposed when the obj is.

+ Sln set up
+ Added Obleak.Fody project - The weaver
+ Added Obleak.Fody.Core - Attribute and HandleWith extension method for IDisposables
+ Unit tests with Class, Method and Constructor level attribute
+ Included a static local weaver which can be used for debugging the weaving process
+ Added Obleak.Fody.nuspec for appveyor deployment of NuGet package

rjlohan commented on Obleak.Fody/Extensions.cs:
```diff
@@ -0,0 +1,91 @@
+    }
+
+    /// <summary>
+    /// Extension methods to reduce some of the bioler plate when working with Mono.Cecil.
```
typo in comment  'bioler'

Syrus666 replied:
Done. Althought expecting a developer to be able to English is ascking a lot.

rjlohan commented on Obleak.Fody/Extensions.cs:
```diff
@@ -0,0 +1,91 @@
+
+namespace Obleak.Fody
+{
+    internal static class Extensions
```
personally prefer classes in own file, but I don't think I have the authority to mandate coding style in this case :laughing:

Syrus666 replied:
I agree up to a point - it's annoying enough not having this as part of the .Net framework (I understand their justification), the idea of having this sole extension method in another file would make me :rage4: 

If there were to ever be another extension method added to Extensions I'd move it out to a separate .cs (originally I had this in the .Core project in it's own file)

rjlohan replied:
Ya no dramas. Also one of the first extensions I write for every project I use which does LINQ. :tired_face:

Syrus666 replied:
Telerik.Windows.Diagrams.Core hehe ;D

rjlohan commented on Obleak.Fody/Extensions.cs:
```diff
@@ -0,0 +1,91 @@
+        /// <summary>
+        /// Calculate if the type, or any of it's base types implement the IDiposable interface
+        /// </summary>
+        internal static bool IsDisposable(this TypeDefinition type)
```
There's a logical inconsistency in this func vs 'GetDisposeMethod'. This one required that type implements IDisposable. GetDisposeMethod only requires an argument-free function with the name 'Dispose'. Might be worth picking one rule (probably this one), which would mean changing GetDisposeMethod to query only against the IDisposable type implementations?

Syrus666 replied:
Thoughts on throwing in GetDisposeMehod (which I've fixed up the name) if the type isn't disposable? 

In what is in effect a script if we've called GetDisposeMehod on something that isn't implementing IDisposable then we're already fucked

rjlohan replied:
I guess all I want is clarity for how this is supposed to work when someone (incorrectly) implements a method called 'Dispose' without the IDisposable interface. Don't think it hurts to add a call to IsDisposable in GetDisposeMethod (or in some other way adjust this method chain to exclude incorrectly implemented Dispose functions)

rjlohan commented on Obleak.Fody/ModuleWeaver.cs:
```diff
@@ -0,0 +1,27 @@
+
+namespace Obleak.Fody
+{
+    public class ModuleWeaver
```
Does this type do something magical which isn't apparent from its lack of usage elsewhere in the code?

Syrus666 replied:
This is what Fody itself uses to do the weave - it's standard across all fody extensions

rjlohan commented on Obleak.Fody/ObleakWeaver.cs:
```diff
@@ -0,0 +1,183 @@
+            LogInfo("Obleak weaving");
+
+            // Validate we have the dependencies we need
+            var rxCore = ModuleDefinition.AssemblyReferences.Where(x => x.Name == "System.Reactive.Core").OrderByDescending(x => x.Version).FirstOrDefault();
```
Is it worth pulling out this to a func? The logic for 'get latest version of named reference or null' is repeated thrice in this method. (Not fussed either way, just highlighting thought bubbles)

Syrus666 replied:
Like it, will do

Commit:
Updated post review feedback

+ Fixed typo in comment
+ Fixed typo in GetDisposeMehod (*Method)
+ Added an extension method to find the latest version of an assembly by name, and optionally invoke an action if not found
+ Validate that the type or base type implements IDisposable before getting the Dispose method. If the type is not disposable, throw an exception
+ Updated .Core assembly info with a small description and author

Syrus666 commented:
Ready for re-review: @rjlohan

rjlohan commented on Obleak.Fody.Core/Properties/AssemblyInfo.cs:
```diff
@@ -0,0 +1,36 @@
+[assembly: AssemblyConfiguration("")]
+[assembly: AssemblyCompany("")]
+[assembly: AssemblyProduct("Obleak.Fody.Core")]
+[assembly: AssemblyCopyright("Copyright ©  2016")]
```
Since this binary eventually deploys with NVivo client, might be worth adding some identifying details to this AssemblyInfo ? (I dunno what though)
Perhaps even a link back to the source repo in the AssemblyDescription, in case we lose memory of where this came from down the track?

rjlohan commented:
Last query on the AssemblyInfo.cs, else all good. :+1:

Syrus666 commented:
Addressed the assembly info with a brief description and author

satsuper commented on .gitignore:
```diff
@@ -0,0 +1,13 @@
+packages/
```
Could just use one of the template gitignores for C# projects here but not fussed

satsuper commented:
:+1: :+1:
Love me some IL
